packages/flutter/lib/src/widgets/selection.dart

```diff
@@ -0,0 +1,977 @@
+            newIndex -= 1;
+            finalResult = SelectionResult.end;
+          } else if (newIndex == 0) {
+            finalResult = currentSelectableResult;
+          } else {
+            forward = false;
+            newIndex -= 1;
+          }
+          break;
+      }
+    }
+    if (isEnd) {
+      currentSelectionEndIndex = newIndex;
+    } else {
+      currentSelectionStartIndex = newIndex;
+    }
+    return finalResult!;
+  }
+}
+
+class _RenderSelectableContainerWidget extends RenderProxyBox with ChangeNotifier, Selectable, SelectionRegisteree {
+  _RenderSelectableContainerWidget({
+    required SelectionContainerDelegate? delegate,
+    required this.containerHasSelectableContent,
+    required SelectionRegistrar? selectionRegistrar,
+    RenderBox? child,
+  }) : _delegate = delegate,
+       super(child) {
+    _delegate?.addListener(notifyListeners);
+    containerHasSelectableContent.addListener(_handleContainerHasSelectableContentChange);
+    _handleContainerHasSelectableContentChange();
+    registrar = selectionRegistrar;
+  }
+
+  SelectionContainerDelegate? get delegate => _delegate;
+  SelectionContainerDelegate? _delegate;
+  set delegate(SelectionContainerDelegate? value) {
+    if (_delegate == null)
+      return;
+    _delegate?.removeListener(notifyListeners);
+    _delegate = value;
+    _delegate?.addListener(notifyListeners);
+  }
+
+  final ValueListenable<bool> containerHasSelectableContent;
+
+  void _handleContainerHasSelectableContentChange() {
+    hasSelectableContent.value = containerHasSelectableContent.value;
+  }
+
+  @override
+  void dispose() {
+    containerHasSelectableContent.removeListener(_handleContainerHasSelectableContentChange);
+    super.dispose();
+  }
+
+  @override
+  void pushHandleLayers(LayerLink? startHandle, LayerLink? endHandle) {
+    _delegate!.pushHandleLayers(startHandle, endHandle);
+  }
+
+  @override
+  Object? getSelectedContent() => _delegate!.getSelectedContent();
+
+  @override
+  SelectionResult dispatchSelectionEvent(SelectionEvent event) {
+    return _delegate!.dispatchSelectionEvent(event);
+  }
+
+  @override
+  SelectionGeometry? get value => _delegate!.value;
+}
+
+class _SelectionRegistrarScope extends InheritedWidget {
+  const _SelectionRegistrarScope({
+    Key? key,
+    required this.registrar,
+    required this.enabled,
+    required Widget child,
+  }) : super(key: key, child: child);
+
+  /// The registrar.
+  final SelectionRegistrar registrar;
+
+  /// whether this enabled
+  final bool enabled;
+
+  @override
+  bool updateShouldNotify(_SelectionRegistrarScope oldWidget) {
+    return oldWidget.registrar != registrar ||
+           oldWidget.enabled != enabled;
+  }
+}
+
+/// The adapter to handles the selections from the [SelectionArea] for
+/// [RichText].
+///
+/// By Wrapping a [RichText] with this widget, the [RichText] will be
+/// selectable under a [SelectionArea].
+///
+/// The [child] must be a [RichText].
```

Comment:
If it must be a RichText, why is it typed as Widget? But also first bullet of: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-interleaving-multiple-concepts-together